src/main/java/io/opentelemetry/semconv/AttributeKeyTemplate.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.semconv;
+
+import io.opentelemetry.api.common.AttributeKey;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.Function;
+
+/**
+ * This class provides a handle for creating and caching dynamic / template-type attributes of the
+ * form <b>&lt;prefix&gt;.&lt;key&gt;</b>. The &lt;prefix&gt; is fixed for a template instance while
+ * {@link AttributeKey}s can be created and are cached for different values of the &lt;key&gt; part.
+ *
+ * <p>An example template-type attribute is the set of attributes for HTTP headers:
+ * <b>http.request.header.&lt;key&gt;</b>
+ *
+ * @param <T> The type of the nested {@link AttributeKey}s.
+ */
+public final class AttributeKeyTemplate<T> {
+
+  private final String prefix;
+  private final Function<String, AttributeKey<T>> keyBuilder;
+  private final ConcurrentMap<String, AttributeKey<T>> keysCache = new ConcurrentHashMap<>(1);
+
+  AttributeKeyTemplate(String prefix, Function<String, AttributeKey<T>> keyBuilder) {
+    this.prefix = prefix;
+    this.keyBuilder = keyBuilder;
+  }
+
+  static AttributeKeyTemplate<String> stringKeyTemplate(String prefix) {
```

Comment:
```suggestion
  public static AttributeKeyTemplate<String> stringKeyTemplate(String prefix) {
```
What do folks think about making these public? As is, they can only be used for attributes in `ResourceAttributes` and `SemanticAttributes`, but its reasonable for a user to want to define their own custom attribute template. Maybe that suggests that ultimately this class should be promoted to `io.opentelemetry.api.common` to live alongside `AttributeKey` and `Attributes`.